M3 lz 34

Eternalynx opened the pull request:
Added highlights for mockup surveys. To create them, click Add Highlight button under the question, then click move mouse click. To save click the Save button (bottom right corner), to cancel click Cancel. To see highlights click the Toggle Highlights button in MockupSurvey view. Click again to hide.

Eternalynx committed:
Added Highlight support for mockup surveys

Eternalynx committed:
Minor text changes

Eternalynx committed:
Minor button changes

Eternalynx committed:
resolving conflicts with develop

Eternalynx committed:
removed unused variables

Eternalynx committed:
removed unused imports

Eternalynx committed:
added question numbers to highlights

Eternalynx committed:
added question numbers to highlights and changed placement of highlight button

agsta commented:
@Eternalynx can you switch the master branch to develop?

agsta commented:
- Currently, the highlights are saved with `mockupId `+ `questionId`, but not with `surveyId `- as a result in a survey view all highlights created for mockup are visible.
- The 'Save' / 'Cancel' buttons could be somewhere higher, they are not always visible for smaller screens
- Maybe we could allow user to set the height/width while uploading the mockup? This way the iframe could be set to fit the content and we wouldn't have the scrollbar issue (however this wouldn't solve the problem with changing pages in the prototype). But it's just a suggestion, we probably don't have time for that

mimagiera commented:
> * Currently, the highlights are saved with `mockupId `+ `questionId`, but not with `surveyId `- as a result in a survey view all highlights created for mockup are visible.
> * The 'Save' / 'Cancel' buttons could be somewhere higher, they are not always visible for smaller screens
> * Maybe we could allow user to set the height/width while uploading the mockup? This way the iframe could be set to fit the content and we wouldn't have the scrollbar issue (however this wouldn't solve the problem with changing pages in the prototype). But it's just a suggestion, we probably don't have time for that

2.  'Save' / 'Cancel' buttons are overlapping with other content, maybe placi it next yo `Add highlight` button
![image](https://user-images.githubusercontent.com/43969709/84692693-91fbef00-af46-11ea-845d-1ae888a3fcb8.png)
3. We could also always set height of iframe to "max" of mockup and probably never have problem with scrolling it. But as you said, another problems will sill exist and we don't have time

mimagiera commented:
Another thing, this text and button should be alligned vertically (or maybe better this text should be a tooltip as @agsta suggests)
![image](https://user-images.githubusercontent.com/43969709/84693099-3847f480-af47-11ea-93ea-bd29a1591de8.png)

mimagiera commented on front/src/components/AddHighlightPopup.js:
```diff
@@ -0,0 +1,148 @@
+            isDrawing: false,
+            posX: 0,
+            posY: 0,
+            hght: 0,
```
why not `height`?

mimagiera commented on front/src/components/AddMockupSurvey.js:
```diff
@@ -306,6 +328,13 @@ class AddMockupSurvey extends Component {
                                                                 </Input>
                                                             </Col>
+                                                            <Col>
+                                                                Highlight doesn't work with scrolling mockups
```
this text could be a const, don't repeat it 4 times

agsta commented:
> Another thing, this text and button should be alligned vertically (or maybe better this text should be a tooltip as @agsta suggests)
> ![image](https://user-images.githubusercontent.com/43969709/84693099-3847f480-af47-11ea-93ea-bd29a1591de8.png)

See: https://reactstrap.github.io/components/tooltips/

mimagiera commented on front/src/components/AddMockupSurvey.js:
```diff
@@ -339,6 +368,13 @@ class AddMockupSurvey extends Component {
                                                                     className="btn btn-primary btn-margin-top">Add Statement
                                                                 </Button>
                                                             </Col>
+                                                            <Col>
```
actually creating this column could be extracted to a function. e.g. (maybe with better function name):
```
    getHighlightCol(index) {
        return <Col>
            Highlight doesn't work with scrolling mockups
            <Button
                onClick={(event) => this.addHighlight(event, index)}
                className="btn btn-primary btn-margin-top">Add Highlight
            </Button>
        </Col>;
    }
```

mimagiera commented on front/src/components/HighlightPopup.js:
```diff
@@ -0,0 +1,82 @@
+import {Layer, Stage, Rect, Group, Text} from "react-konva";
+
+
+
```
sooo many empty lines

mimagiera commented on front/src/components/HighlightPopup.js:
```diff
@@ -0,0 +1,82 @@
+            highlights: []
+        };
+
+
```
sooo many empty lines

mimagiera commented on src/main/java/lz/modelTesting/controllers/HighlightsController.java:
```diff
@@ -0,0 +1,69 @@
+package lz.modelTesting.controllers;
+
+
```
sooo many empty lines

mimagiera commented on src/main/java/lz/modelTesting/controllers/HighlightsController.java:
```diff
@@ -0,0 +1,69 @@
+@CrossOrigin(origins = "*", allowedHeaders = "*")
+public class HighlightsController {
+
+    private final static String MOCKUP_ID = "mockupId";
```
that could be reused from `SurveysController` and even better extracted to another place (another class) and be used here and in `SurveysController` from that another place

mimagiera commented on src/main/java/lz/modelTesting/controllers/HighlightsController.java:
```diff
@@ -0,0 +1,69 @@
+        String posX = request.getParameter(POS_X);
+        String posY = request.getParameter(POS_Y);
+
+
```
sooo many empty lines

mimagiera commented on src/main/java/lz/modelTesting/documents/Highlight.java:
```diff
@@ -0,0 +1,32 @@
+    private String posY;
+
+
+
```
sooo many empty lines

mimagiera commented on src/main/java/lz/modelTesting/controllers/HighlightsController.java:
```diff
@@ -0,0 +1,69 @@
+@RestController
+@RequestMapping("/highlights")
+@CrossOrigin(origins = "*", allowedHeaders = "*")
+public class HighlightsController {
```
as @agsta said, highlights should be also taken by `surveyId`

mimagiera reviewed (changes requested):
* pls format code properly (ctrl alt L) and imports (ctrl alt O)
* delete all unnecessary empty lines
* change highlight entity to also contain surveyId
* make another name refactors / extract redundant code to functions

Eternalynx commented:
Problem with changing mockupId to surveyId - at the point of creating highlight surveyId doesn't exist because survey isn't saved yet. Should I try to manually create surveyId before? But how would I do that?

Eternalynx commented:
Also, can I place save and cancel buttons on the bar next to fullscreen and show preview? I cannot find any other place that wouldn't interfere with anything

mimagiera commented:
> Problem with changing mockupId to surveyId - at the point of creating highlight surveyId doesn't exist because survey isn't saved yet. Should I try to manually create surveyId before? But how would I do that?

How about setting surveyId in highlight during saving whole survey to database?

mimagiera commented:
> Also, can I place save and cancel buttons on the bar next to fullscreen and show preview? I cannot find any other place that wouldn't interfere with anything

MAybe next to button "Add highlight"?

Eternalynx commented:
There are multiple add highlight buttons with different positions depending on question type

mimagiera commented:
> There are multiple add highlight buttons with different positions depending on question type

maybe dynamically create those buttons next to this "add highlight" which was clicked

Eternalynx commented:
> > There are multiple add highlight buttons with different positions depending on question type
> 
> maybe dynamically create those buttons next to this "add highlight" which was clicked

I have no idea how to get position of clicked button



> > Problem with changing mockupId to surveyId - at the point of creating highlight surveyId doesn't exist because survey isn't saved yet. Should I try to manually create surveyId before? But how would I do that?
> 
> How about setting surveyId in highlight during saving whole survey to database?

Should I pass to the backend list of highlight ids then? How about procreating a survey in componentDidMount and updating it on submit? Would that work?

agsta committed:
[M3] tooltip added

Eternalynx committed:
added surveyId to highlight but had to remove handleRedirect from AddMockupSurvey due to the timing of axios calls.

Changed place of Close and Save buttons in addHighlightPopup. Right now above the survey

Eternalynx committed:
merged M3-fix
